src/components/on-this-page/helpers.js

```diff
@@ -0,0 +1,108 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import Icon from '@mapbox/mr-ui/icon';
+
+export function findSubItems(pickedHeadings, index) {
+  const h2Indices = pickedHeadings.reduce((a, item, i) => {
+    if (item.level === 2) a.push(i);
+    return a;
+  }, []);
+  const nextH2 = h2Indices[h2Indices.indexOf(index) + 1];
+  return pickedHeadings.slice(index, nextH2).filter((f) => f.level === 3);
+}
+
+// splits an array of headings into a nested array
+export function buildSections(headings) {
+  if (!headings) return [];
+  const pickedHeadings = headings.filter(
+    (heading) => heading.level === 2 || heading.level === 3
+  );
+  return pickedHeadings.reduce((arr, heading, index) => {
+    if (heading.level === 2) {
+      arr.push({
+        id: heading.slug,
+        value: heading.text,
+        children: findSubItems(pickedHeadings, index).map((sub) => ({
+          id: sub.slug,
+          value: sub.text,
+          ...(sub.icon && { icon: sub.icon })
+        }))
+      });
+    }
+    return arr;
+  }, []);
+}
+
+// finds the active header anchor
+export function getActiveHeaderAnchor(topOffset) {
```

Comment:
This function was heavily borrowed from https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-theme-classic/src/theme/hooks/useTOCHighlight.ts